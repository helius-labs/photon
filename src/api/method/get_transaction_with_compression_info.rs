use crate::common::typedefs::account::Account;
use crate::common::typedefs::serializable_signature::SerializableSignature;
use crate::common::typedefs::token_data::TokenData;
use crate::ingester::parser::parse_transaction;
use crate::ingester::persist::parse_token_data;

use sea_orm::DatabaseConnection;
use serde::{Deserialize, Serialize};
use solana_client::nonblocking::rpc_client::RpcClient;
use solana_client::rpc_config::RpcTransactionConfig;
use solana_client::rpc_request::RpcRequest;
use solana_sdk::commitment_config::{CommitmentConfig, CommitmentLevel};
use solana_transaction_status::{EncodedConfirmedTransactionWithStatusMeta, UiTransactionEncoding};
use utoipa::{
    openapi::{ObjectBuilder, RefOr, Schema, SchemaType},
    ToSchema,
};

use super::{
    super::error::PhotonApiError, get_multiple_compressed_accounts::fetch_accounts_from_hashes,
    utils::parse_account_model,
};

const RPC_CONFIG: RpcTransactionConfig = RpcTransactionConfig {
    encoding: Some(UiTransactionEncoding::Base64),
    commitment: Some(CommitmentConfig {
        commitment: CommitmentLevel::Confirmed,
    }),
    max_supported_transaction_version: Some(0),
};

// We do not use generics to simply documentation generation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct GetTransactionRequest {
    pub signature: SerializableSignature,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, ToSchema)]
pub struct CompressionInfo {
    pub closed_accounts: Vec<AccountWithOptionalTokenData>,
    pub opened_accounts: Vec<AccountWithOptionalTokenData>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, ToSchema)]
pub struct AccountWithOptionalTokenData {
    pub account: Account,
    pub optional_token_data: Option<TokenData>,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct GetTransactionResponse {
    pub transaction: EncodedConfirmedTransactionWithStatusMeta,
    pub compression_info: CompressionInfo,
}

impl<'__s> ToSchema<'__s> for GetTransactionResponse {
    fn schema() -> (&'__s str, RefOr<Schema>) {
        let schema = Schema::Object(
            ObjectBuilder::new()
                .schema_type(SchemaType::Object)
                .description(Some(
                    "A Solana transaction with additional compression information",
                ))
                // TODO: Improve OpenAPI documentation here.
                .property(
                    "transaction",
                    ObjectBuilder::new()
                        .schema_type(SchemaType::Object)
                        .description(Some(
                            "An encoded confirmed transaction with status meta".to_string(),
                        ))
                        .build(),
                )
                .property("compression_info", CompressionInfo::schema().1)
                .build(),
        );

        ("GetTransactionResponse", RefOr::T(schema))
    }

    fn aliases() -> Vec<(&'static str, utoipa::openapi::schema::Schema)> {
        Vec::new()
    }
}
fn parse_optional_token_data(
    account: Account,
) -> Result<AccountWithOptionalTokenData, PhotonApiError> {
    let hash = account.hash.clone();
    Ok(AccountWithOptionalTokenData {
        optional_token_data: parse_token_data(&account).map_err(|e| {
            PhotonApiError::UnexpectedError(format!(
                "Failed to parse token data for account {}: {}",
                hash, e
            ))
        })?,
        account,
    })
}

fn parse_optional_token_data_for_multiple_accounts(
    accounts: Vec<Account>,
) -> Result<Vec<AccountWithOptionalTokenData>, PhotonApiError> {
    accounts
        .into_iter()
        .map(parse_optional_token_data)
        .collect()
}

fn clone_tx(
    txn: &EncodedConfirmedTransactionWithStatusMeta,
) -> EncodedConfirmedTransactionWithStatusMeta {
    EncodedConfirmedTransactionWithStatusMeta {
        slot: txn.slot,
        transaction: txn.transaction.clone(),
        block_time: txn.block_time,
    }
}

pub async fn get_transaction_helper(
    conn: &DatabaseConnection,
    signature: SerializableSignature,
    txn: EncodedConfirmedTransactionWithStatusMeta,
) -> Result<GetTransactionResponse, PhotonApiError> {
    // Ignore if tx failed or meta is missed
    let meta = txn.transaction.meta.as_ref();
    if meta.map(|meta| meta.status.is_err()).unwrap_or(true) {
        return Err(PhotonApiError::ValidationError(
            "Transaction missing metatada information".to_string(),
        ));
    }
    let slot = txn.slot;

    let status_update = parse_transaction(
        &clone_tx(&txn).try_into().map_err(|_e| {
            PhotonApiError::UnexpectedError(format!("Failed to parse transaction {}", signature.0))
        })?,
        slot,
    )
    .map_err(|_e| {
        PhotonApiError::UnexpectedError(format!("Failed to parse transaction {}", signature.0))
    })?;

    let closed_accounts = fetch_accounts_from_hashes(
        conn,
        status_update.in_accounts.iter().cloned().collect(),
        true,
    )
    .await?
    .into_iter()
    .map(parse_account_model)
    .collect::<Result<Vec<Account>, PhotonApiError>>()?;

    Ok(GetTransactionResponse {
        transaction: txn,
        compression_info: CompressionInfo {
            closed_accounts: parse_optional_token_data_for_multiple_accounts(closed_accounts)?,
            opened_accounts: parse_optional_token_data_for_multiple_accounts(
                status_update.out_accounts,
            )?,
        },
    })
}

pub async fn get_transaction_with_compression_info(
    conn: &DatabaseConnection,
    rpc_client: &RpcClient,
    request: GetTransactionRequest,
) -> Result<GetTransactionResponse, PhotonApiError> {
    let txn: EncodedConfirmedTransactionWithStatusMeta = rpc_client
        .send(
            RpcRequest::GetTransaction,
            serde_json::json!([request.signature.0.to_string(), RPC_CONFIG,]),
        )
        .await
        .map_err(|e| {
            PhotonApiError::UnexpectedError(format!(
                "Failed to fetch transaction {}: {}",
                request.signature.0, e
            ))
        })?;
    get_transaction_helper(conn, request.signature, txn).await
}
