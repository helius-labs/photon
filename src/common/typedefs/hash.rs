use core::fmt;
use std::mem;

use serde::{Deserialize, Serialize};
use utoipa::openapi::{ObjectBuilder, RefOr, Schema, SchemaType};
use utoipa::ToSchema;

use serde::de::{self, Visitor};
use serde::ser::Serializer;
use serde::Deserializer;
#[allow(unused_imports)]
use solana_sdk::pubkey::Pubkey;
use thiserror::Error;

// Maximum length of a 32 byte base58 encoded hash
const MAX_BASE58_LEN: usize = 44;

/// `Hash` is a struct that represents a 32-byte hash.
///
/// This is similar to a Solana Public Key, which is also a 32-byte hash. However, while a Solana
/// Public Key is specifically an ed25519 public key, a `Hash` is not necessarily an ed25519 hash.
///
/// Also a `Hash` is used to represent other types of hashes, such as account hashes and Merkle
/// tree node hashes. This is why `Hash` is a separate struct from the the Solana Public Key struct.
#[derive(Default, Clone, PartialEq, Eq)]
pub struct Hash([u8; 32]);

impl Hash {
    pub fn to_vec(&self) -> Vec<u8> {
        self.0.to_vec()
    }

    pub fn to_base58(&self) -> String {
        bs58::encode(self.0).into_string()
    }

    pub fn new_unique() -> Self {
        // Slightly hacky way to get 32 random bytes
        Hash(Pubkey::new_unique().to_bytes())
    }
}

impl<'__s> ToSchema<'__s> for Hash {
    fn schema() -> (&'__s str, RefOr<Schema>) {
        let schema = Schema::Object(
            ObjectBuilder::new()
                .schema_type(SchemaType::String)
                .description(Some("A 32-byte hash represented as a base58 string."))
                .example(Some(serde_json::Value::String(
                    Hash::new_unique().to_base58(),
                )))
                .build(),
        );

        ("Hash", RefOr::T(schema))
    }

    fn aliases() -> Vec<(&'static str, utoipa::openapi::schema::Schema)> {
        Vec::new()
    }
}

#[derive(Error, Debug, Serialize, Clone, PartialEq, Eq)]
pub enum ParseHashError {
    #[error("String is the wrong size")]
    WrongSize,
    #[error("Invalid hash input")]
    Invalid,
}

impl TryFrom<&str> for Hash {
    type Error = ParseHashError;

    fn try_from(s: &str) -> Result<Self, Self::Error> {
        if s.len() > MAX_BASE58_LEN {
            return Err(ParseHashError::WrongSize);
        }
        let bytes = bs58::decode(s)
            .into_vec()
            .map_err(|_| ParseHashError::Invalid)?;

        bytes.try_into()
    }
}

impl From<Hash> for Vec<u8> {
    fn from(val: Hash) -> Self {
        val.0.to_vec()
    }
}

impl From<[u8; 32]> for Hash {
    fn from(bytes: [u8; 32]) -> Self {
        Hash(bytes)
    }
}

impl From<Hash> for [u8; 32] {
    fn from(val: Hash) -> Self {
        val.0
    }
}

impl TryFrom<Vec<u8>> for Hash {
    type Error = ParseHashError;

    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
        if bytes.len() != mem::size_of::<Hash>() {
            Err(ParseHashError::WrongSize)
        } else {
            let bytes: [u8; 32] = bytes.try_into().map_err(|_| ParseHashError::Invalid)?;
            Ok(Hash(bytes))
        }
    }
}

impl std::fmt::Display for Hash {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_base58())
    }
}

impl From<Hash> for String {
    fn from(val: Hash) -> Self {
        val.to_base58()
    }
}

impl fmt::Debug for Hash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Hash({})", self.to_base58())
    }
}

struct Base58VisitorHash;

impl<'de> Visitor<'de> for Base58VisitorHash {
    type Value = Hash;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("a base58 encoded string")
    }

    fn visit_str<E: de::Error>(self, value: &str) -> Result<Self::Value, E> {
        Hash::try_from(value).map_err(|e| E::custom(e.to_string()))
    }
}

impl<'de> Deserialize<'de> for Hash {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        deserializer.deserialize_str(Base58VisitorHash)
    }
}

impl Serialize for Hash {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let base58_string = bs58::encode(self.0).into_string();
        serializer.serialize_str(&base58_string)
    }
}

#[test]
fn test_serialization() {
    // Hacky way to get 32 bytes
    let hash = Hash(Pubkey::new_unique().to_bytes());
    let serialized = serde_json::to_string(&hash).unwrap();
    let deserialized: Hash = serde_json::from_str(&serialized).unwrap();
    assert_eq!(hash, deserialized);
}
