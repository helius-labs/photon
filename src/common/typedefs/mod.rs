pub mod hash;
pub mod serializable_pubkey;
