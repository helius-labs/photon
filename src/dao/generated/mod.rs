//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.6

pub mod prelude;

pub mod account_transactions;
pub mod accounts;
pub mod blocks;
pub mod owner_balances;
pub mod state_trees;
pub mod token_accounts;
pub mod token_owner_balances;
pub mod transactions;
