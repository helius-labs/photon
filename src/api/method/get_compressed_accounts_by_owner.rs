use crate::dao::generated::accounts;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QueryOrder, QuerySelect};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{
    super::error::PhotonApiError,
    utils::{Context, Limit, PAGE_LIMIT},
};
use crate::common::typedefs::{hash::Hash, serializable_pubkey::SerializablePubkey};

use super::utils::{parse_account_model, Account};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Options {
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedAccountsByOwnerRequest {
    pub owner: SerializablePubkey,
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct PaginatedAccountList {
    pub items: Vec<Account>,
    pub cursor: Option<Hash>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct GetCompressedAccountsByOwnerResponse {
    pub context: Context,
    pub value: PaginatedAccountList,
}

pub async fn get_compressed_accounts_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedAccountsByOwnerRequest,
) -> Result<GetCompressedAccountsByOwnerResponse, PhotonApiError> {
    let context = Context::extract(conn).await?;
    let GetCompressedAccountsByOwnerRequest {
        owner,
        cursor,
        limit,
    } = request;

    let mut filter = accounts::Column::Owner
        .eq::<Vec<u8>>(owner.into())
        .and(accounts::Column::Spent.eq(false));

    if let Some(cursor) = cursor {
        filter = filter.and(accounts::Column::Hash.gt::<Vec<u8>>(cursor.into()));
    }
    let mut query_limit = PAGE_LIMIT;
    if let Some(limit) = limit {
        query_limit = limit.value();
    }

    let result = accounts::Entity::find()
        .order_by(accounts::Column::Hash, sea_orm::Order::Asc)
        .limit(query_limit)
        .filter(filter)
        .all(conn)
        .await?;

    let items = result
        .into_iter()
        .map(parse_account_model)
        .collect::<Result<Vec<Account>, PhotonApiError>>()?;

    let mut cursor = items.last().map(|u| u.hash.clone());
    if items.len() < query_limit as usize {
        cursor = None;
    }

    Ok(GetCompressedAccountsByOwnerResponse {
        context,
        value: PaginatedAccountList { items, cursor },
    })
}
