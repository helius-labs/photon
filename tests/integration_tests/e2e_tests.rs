use function_name::named;
use photon_indexer::api::method::get_compressed_accounts_by_owner::GetCompressedAccountsByOwnerRequest;
use photon_indexer::api::method::get_transaction_with_compression_info::get_transaction_helper;
use photon_indexer::common::typedefs::serializable_pubkey::SerializablePubkey;
use photon_indexer::ingester::index_block;
use solana_sdk::pubkey::Pubkey;

use crate::utils::*;
use insta::assert_json_snapshot;
use photon_indexer::api::method::utils::GetCompressedTokenAccountsByOwner;
use photon_indexer::common::typedefs::hash::Hash;
use photon_indexer::dao::generated::blocks;
use photon_indexer::ingester::typedefs::block_info::{BlockInfo, BlockMetadata};
use sea_orm::ColumnTrait;
use sea_orm::EntityTrait;
use sea_orm::QueryFilter;
use serial_test::serial;

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_mint_and_transfer(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use std::str::FromStr;

    
    use photon_indexer::{
        api::method::{
            get_compression_signatures_for_token_owner::GetCompressionSignaturesForTokenOwnerRequest,
            get_multiple_compressed_account_proofs::HashList, utils::Limit,
        },
        common::typedefs::serializable_signature::SerializableSignature,
    };
    use solana_sdk::signature::Signature;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let payer_pubkey =
        SerializablePubkey::try_from("Fg5TtbkvcqLedmwFqqbWty9NRqMKKMqy9zNLZZRQXSDc").unwrap();
    let bob_pubkey =
        SerializablePubkey::try_from("4rANawp9ahHmvaVi3CCeJSam8JTA4DDY3htJKnhYxbHN").unwrap();
    let charles_pubkey =
        SerializablePubkey::try_from("GyXNDaUUBwzYp24vEzang6SyCzcA8L3csZkiALqF6KRs").unwrap();

    let mint_tx =
        "2DcgLz1KwPaMoiRAAmYAirx4213swLbaSM1YYuCkbgzngq61XSWaFCRGuFoMgqW5htsEZRKBm6uQQPAPCXUVL8eH";
    let transfer_tx =
        "DFE7shhwCWd5P14NfA1ELSpd96FaJhBkjSet2Y7t7CtM4oLhP17AYKtyMBjZWi4hgFKvqET8zcAQUDuAC6sARfk";
    let transfer_txn2: &str =
        "5FDV6xrPRyfEUP2Kg7mwZVj34LVEfq9gsrquyFkvmmdPiiEn3Bhd8sTJTe6UhszNHLgwWfV6j9X4rsceL9mNNyRZ";
    let transfer_txn3 =
        "2pGQ7MgWuhrKpeSAzs58CXttKDiws33X68hQv4tRm6NUH6JCsCBk5uteNcjt2fXdToYhC6s8b6SDzoXV9CLJTo94";

    let txs = [mint_tx, transfer_tx, transfer_txn2, transfer_txn3];

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    for tx in txs {
        index_transaction(&setup, tx).await;
    }
    for (person, pubkey) in [
        ("bob", bob_pubkey.clone()),
        ("charles", charles_pubkey.clone()),
        ("payer", payer_pubkey.clone()),
    ] {
        let accounts = setup
            .api
            .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
                owner: pubkey.clone(),
                ..Default::default()
            })
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-accounts", name.clone(), person), accounts);
        let proofs = setup
            .api
            .get_multiple_compressed_account_proofs(HashList(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.account.hash.clone())
                    .collect(),
            ))
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), person), proofs);

        let mut cursor = None;
        let limit = Limit::new(1).unwrap();
        let mut signatures = Vec::new();
        loop {
            let res = setup
                .api
                .get_compression_signatures_for_token_owner(
                    GetCompressionSignaturesForTokenOwnerRequest {
                        owner: pubkey.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    },
                )
                .await
                .unwrap()
                .value;
            signatures.extend(res.items);
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_json_snapshot!(
            format!("{}-{}-token-signatures", name.clone(), person),
            signatures
        );

        let token_balances = setup
                .api
                .get_compressed_token_balances_by_owner(photon_indexer::api::method::get_compressed_token_balances_by_owner::GetCompressedTokenBalancesByOwnerRequest {
                    owner: pubkey.clone(),
                    ..Default::default()
                })
                .await
                .unwrap();

        assert_json_snapshot!(
            format!("{}-{}-token-balances", name.clone(), person),
            token_balances
        );
    }
    for (txn_name, txn_signature) in [("mint", mint_tx), ("transfer", transfer_tx)] {
        let txn = cached_fetch_transaction(&setup, txn_signature).await;
        let txn_signature = SerializableSignature(Signature::from_str(txn_signature).unwrap());
        // Test get transaction
        let parsed_transaction: photon_indexer::api::method::get_transaction_with_compression_info::GetTransactionResponse = get_transaction_helper(&setup.db_conn, txn_signature, txn).await.unwrap();
        assert_json_snapshot!(
            format!("{}-{}-transaction", name.clone(), txn_name),
            parsed_transaction
        );
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_lamport_transfers(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use photon_indexer::api::method::get_multiple_compressed_account_proofs::HashList;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let compress_tx =
        "3ZmBETHr8tHPr7iKdgW4NjTzcUCUHtSBFQD2JwSniShBV9fix6xb6FxRyR1cMRW83WerRYU5b1uxDLM3GqomCkYx";
    let transfer_tx1 =
        "64tDNpAnQHZsuiH8UG8vXDzXbwGHdghfLYerPdaVdgkQ2kSTE65kDT7ajzzhd2ZDW8nVoKYztnjxCaV8bWJxNCXR";
    let transfer_tx2 =
        "yK3j6TkABz2SVBtvsVwJV5ZoLv1tJBrBR8Z8JsosHQGbeizt7yx8Qx1Nva1ytcKa7hAXsT7R6WyQUiCH43Cwtk5";

    let payer_pubkey = SerializablePubkey(
        Pubkey::try_from("8wSvtG3dvL7Nso8E3wx6L9SwTRiZuR8ojRX5D6mwx7hJ").unwrap(),
    );
    let receiver_pubkey = SerializablePubkey(
        Pubkey::try_from("G1zCvDhtSdV6m92q3tXBEE4d531fBx1vD7EUpeeQZrNp").unwrap(),
    );
    let txs = [compress_tx, transfer_tx1, transfer_tx2];
    let number_of_indexes = [1, 2];

    for individually in [true, false] {
        reset_tables(&setup.db_conn).await.unwrap();
        for number in number_of_indexes.iter() {
            for _ in 0..*number {
                // HACK: We index a block so that API methods can fetch the current slot.
                index_block(
                    &setup.db_conn,
                    &BlockInfo {
                        metadata: BlockMetadata {
                            slot: 0,
                            ..Default::default()
                        },
                        ..Default::default()
                    },
                )
                .await
                .unwrap();

                match individually {
                    true => {
                        for tx in txs.iter() {
                            index_transaction(&setup, tx).await;
                        }
                    }
                    false => {
                        index_multiple_transactions(&setup, txs.to_vec()).await;
                    }
                }
            }
        }
        for (owner, owner_name) in [
            (payer_pubkey.clone(), "payer"),
            (receiver_pubkey.clone(), "receiver"),
        ] {
            let accounts = setup
                .api
                .get_compressed_accounts_by_owner(GetCompressedAccountsByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();
            assert_json_snapshot!(
                format!("{}-{}-accounts", name.clone(), owner_name),
                accounts
            );
            let proofs = setup
                .api
                .get_multiple_compressed_account_proofs(HashList(
                    accounts
                        .value
                        .items
                        .iter()
                        .map(|x| x.hash.clone())
                        .collect(),
                ))
                .await
                .unwrap();
            assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), owner_name), proofs);

            let signatures = setup
                .api
                .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            let limit = photon_indexer::api::method::utils::Limit::new(1).unwrap();
            let mut cursor = None;
            let mut paginated_signatures = Vec::new();
            loop {
                let res = setup
                    .api
                    .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                        owner: owner.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    })
                    .await
                    .unwrap()
                    .value;
                paginated_signatures.extend(res.items);
                cursor = res.cursor;
                if cursor.is_none() {
                    break;
                }
            }
            assert_eq!(signatures.value.items, paginated_signatures);
            assert_json_snapshot!(
                format!("{}-{}-signatures", owner_name, name.clone()),
                signatures
            );

            let balance = setup
                .api
                .get_compressed_balance_by_owner(photon_indexer::api::method::get_compressed_balance_by_owner::GetCompressedBalanceByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            assert_eq!(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.lamports.0)
                    .sum::<u64>(),
                balance.value.0,
            );
        }
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_index_block_metadata(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Mainnet,
            db_backend,
        },
    )
    .await;

    let slot = 254170887;
    let block = cached_fetch_block(&setup, slot).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    let filter = blocks::Column::Slot.eq(block.metadata.slot);

    let block_model = blocks::Entity::find()
        .filter(filter)
        .one(setup.db_conn.as_ref())
        .await
        .unwrap()
        .unwrap();

    assert_eq!(block_model.slot, 254170887);
    assert_eq!(block_model.parent_slot, 254170886);
    assert_eq!(
        Hash::try_from(block_model.parent_blockhash)
            .unwrap()
            .to_string(),
        "9BMHTdybcGah8PWtCzu8tVFDBXmiEHZZDf3FaZ651Nf"
    );
    assert_eq!(
        Hash::try_from(block_model.blockhash).unwrap().to_string(),
        "5GG5pzTbH6KgZM54M9XWfBNmBupQuZXdQxoZRRQXHcpM"
    );
    assert_eq!(block_model.block_height, 234724352);
    assert_eq!(block_model.block_time, 1710441678);

    // Verify that we don't get an error if we try to index the same block again
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot);

    // Verify that get_indexer_slot() gets updated a new block is indexed.
    let block = cached_fetch_block(&setup, slot + 1).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot + 1);
}
