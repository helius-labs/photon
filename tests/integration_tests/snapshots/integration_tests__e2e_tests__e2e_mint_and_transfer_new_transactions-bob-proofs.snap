---
source: tests/integration_tests/e2e_tests.rs
expression: proofs
---
{
  "context": {
    "slot": 0
  },
  "value": [
    {
      "proof": [
        "11111111111111111111111111111111",
        "2ogtRD2rJqm24o8edvzffTWxXaSD3TTQR8ZVQYhnwNUp",
        "akhmi56gmUvnwAnbb43AgVYcd7612LzpFRfezMLfVVC",
        "2gQksfK543Jw196amPxXWceFzNTfRHiqb7x2edvupftT",
        "Y8pRoGMGXJPGZyb7m4sqkPsTSJvzZ7oSVnaybCcYbgd",
        "3w895nemtSxb1fEjFT3cXkVpW3y6f9uHQn43jEwu1RWc",
        "46JM2e72mdEAxkSeXc65b3aLWkAQpX6NF7mdyuXQKguu",
        "WKLyYL7wpK5zVsWoMU2rcUiMwRWp9PowG83GHHH1dqn",
        "4CztDUcRvGkTyFoVEXXcBwd5UwqJcJJWpdL8FFERrqkU",
        "yjDkqNqZtBwpG7iUwT3hqKNKZAP9CcBDyBBuGPNk2jQ",
        "2r8qWibedWpM7QkW6QvDDUEQAHT9isX629UKcUxbY8BT",
        "38Aoz1ZUAnRqSCxdQuEkLhzYMRi1Tu1rSBy1UymA8yfK",
        "3zBeAgo5EwPrgsXRLUV6DhQrsUUhwbmQXPnZkudhsnSD",
        "2Q5cDvZW6hTxFFDTHhAf129PH4z8EZLYA1meeZpPW76F",
        "2gnsMw1zhpqGeAakoQLarCXVUaeEGDHm5uma1Jznx7qD",
        "3MXZLrn3tnLMWJznkMombRHkgkfsiBTPru2wjHDd3kLh",
        "3rrDm8iED4sDyonpLs9J3TDoTP8U4jsvXCmoupoCvvKc",
        "48YJtP1G5pKWu7Jy1NRagdj29dwxVXFsySrBVw1nALhX",
        "22tjwuiv6VgeMDs8qXpkZ7cY5rcgrFrG5knp7swZ1wQ2",
        "2dS4UCZ6RJvwPBaWmcz1L7ips34xnbNMfK9D8UHdFJfT",
        "3EdF4zvu7uwW7xYdnaKUPzqDY5Msq9FfWn6HgEnRTkDB",
        "2jzv6Z8vEXSYRXQCFnVYvzgJ2RfyDN1QEjdHm6CBvJUg",
        "2fomPS2xUPwXxVF8hx5YtMpBHmQKSZKushUiWYsy9Eqx",
        "3Pi3nRu4fWpnW5rYDAoraX1Q2xDCRQ24Dq8TEY7gaSBn",
        "3dbMMyehykAriNZ8FMJgPc1N6YSzm9ZfV2HcW9sbaXzP",
        "4EE5KQnQ6Tvo78gGhTu8serpJV4srbRhYZ3rdvCiRa5e"
      ],
      "root": "ZMg67iuQAyRUALR4TmAMyWGVEXV2e8yCpRSqdnR8Y1n",
      "leafIndex": 6,
      "hash": "2WfNf8qZoMNGR4mAPX8toADTdRpcKYTX1KUUAEHMi7dJ",
      "merkleTree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
      "rootSeq": 7
    },
    {
      "proof": [
        "1jKDQdFp9seA3B9kbTisBo8BGsMTgXYNiotaDYdcTtB",
        "2NFr4zNFkEpvpK7VJrcWaMKZSw2tuApoNjWTyLFPdihF",
        "3zsuc4AdBg45FebfR1QNQVpPY9US638xanMCNiCZxTKb",
        "2gQksfK543Jw196amPxXWceFzNTfRHiqb7x2edvupftT",
        "Y8pRoGMGXJPGZyb7m4sqkPsTSJvzZ7oSVnaybCcYbgd",
        "3w895nemtSxb1fEjFT3cXkVpW3y6f9uHQn43jEwu1RWc",
        "46JM2e72mdEAxkSeXc65b3aLWkAQpX6NF7mdyuXQKguu",
        "WKLyYL7wpK5zVsWoMU2rcUiMwRWp9PowG83GHHH1dqn",
        "4CztDUcRvGkTyFoVEXXcBwd5UwqJcJJWpdL8FFERrqkU",
        "yjDkqNqZtBwpG7iUwT3hqKNKZAP9CcBDyBBuGPNk2jQ",
        "2r8qWibedWpM7QkW6QvDDUEQAHT9isX629UKcUxbY8BT",
        "38Aoz1ZUAnRqSCxdQuEkLhzYMRi1Tu1rSBy1UymA8yfK",
        "3zBeAgo5EwPrgsXRLUV6DhQrsUUhwbmQXPnZkudhsnSD",
        "2Q5cDvZW6hTxFFDTHhAf129PH4z8EZLYA1meeZpPW76F",
        "2gnsMw1zhpqGeAakoQLarCXVUaeEGDHm5uma1Jznx7qD",
        "3MXZLrn3tnLMWJznkMombRHkgkfsiBTPru2wjHDd3kLh",
        "3rrDm8iED4sDyonpLs9J3TDoTP8U4jsvXCmoupoCvvKc",
        "48YJtP1G5pKWu7Jy1NRagdj29dwxVXFsySrBVw1nALhX",
        "22tjwuiv6VgeMDs8qXpkZ7cY5rcgrFrG5knp7swZ1wQ2",
        "2dS4UCZ6RJvwPBaWmcz1L7ips34xnbNMfK9D8UHdFJfT",
        "3EdF4zvu7uwW7xYdnaKUPzqDY5Msq9FfWn6HgEnRTkDB",
        "2jzv6Z8vEXSYRXQCFnVYvzgJ2RfyDN1QEjdHm6CBvJUg",
        "2fomPS2xUPwXxVF8hx5YtMpBHmQKSZKushUiWYsy9Eqx",
        "3Pi3nRu4fWpnW5rYDAoraX1Q2xDCRQ24Dq8TEY7gaSBn",
        "3dbMMyehykAriNZ8FMJgPc1N6YSzm9ZfV2HcW9sbaXzP",
        "4EE5KQnQ6Tvo78gGhTu8serpJV4srbRhYZ3rdvCiRa5e"
      ],
      "root": "ZMg67iuQAyRUALR4TmAMyWGVEXV2e8yCpRSqdnR8Y1n",
      "leafIndex": 3,
      "hash": "36ewgEJTqSu4enYGWUcxm4CYezR3kgfHBeV29WiEgKaY",
      "merkleTree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
      "rootSeq": 7
    }
  ]
}
